Add more container visualizations

sockeye-d opened the pull request:
This adds editor visualizations for
* GridContainer
* AspectRatioContainer
* CenterContainer
* ScrollContainer
* MarginContainer
* SplitContainer
* FlowContainer
* BoxContainer

https://github.com/user-attachments/assets/12afabfe-2c0a-488e-9836-f756ddcf2fed

I'm not sure if the friend class + exposing private fields for storing sorting data is the right approach, but it seems to work. It may make the containers have slightly worse performance, but I'm not sure how to disable it when the editor plugin isn't compiled.

Closes godotengine/godot-proposals#13753
Closes godotengine/godot-proposals#13742

sockeye-d committed:
Add container visualizations for
* GridContainer
* AspectRatioContainer
* CenterContainer
* ScrollContainer
* MarginContainer
* SplitContainer
* FlowContainer
* BoxContainer

AThousandShips commented on editor/scene/gui/container_editor_plugin.cpp:
```diff
@@ -0,0 +1,290 @@
+
+#include "container_editor_plugin.h"
+
+#include "core/math/math_funcs.h"
```
```suggestion
```
Already included

AThousandShips commented on editor/scene/gui/container_editor_plugin.cpp:
```diff
@@ -0,0 +1,290 @@
+			for (int cell_index = 0; cell_index < grid_container->cell_sizes.size(); cell_index++) {
+				int32_t cell_offset = grid_container->cell_sizes[cell_index];
+				bool draw_row_lines = cell_index >= grid_container->get_columns() - 1;
+				Vector2 p1, p2;
```
```suggestion
				Vector2 p1;
				Vector2 p2;
```
For all, one declaration per line please

AThousandShips commented on editor/scene/gui/container_editor_plugin.cpp:
```diff
@@ -0,0 +1,290 @@
+void ContainerEditorPlugin::edit(Object *p_object) {
+	if (container) {
+		container->disconnect(SNAME("draw"), callable_mp(CanvasItemEditor::get_singleton(), &CanvasItemEditor::update_viewport));
+		container->disconnect(SNAME("sort_children"), callable_mp(CanvasItemEditor::get_singleton(), &CanvasItemEditor::update_viewport));
```
```suggestion
		container->disconnect(SceneStringName(draw), callable_mp(CanvasItemEditor::get_singleton(), &CanvasItemEditor::update_viewport));
		container->disconnect(SceneStringName(sort_children), callable_mp(CanvasItemEditor::get_singleton(), &CanvasItemEditor::update_viewport));
```

AThousandShips commented on editor/scene/gui/container_editor_plugin.cpp:
```diff
@@ -0,0 +1,290 @@
+
+	if (container) {
+		container->connect(SNAME("draw"), callable_mp(CanvasItemEditor::get_singleton(), &CanvasItemEditor::update_viewport));
+		container->connect(SNAME("sort_children"), callable_mp(CanvasItemEditor::get_singleton(), &CanvasItemEditor::update_viewport));
```
```suggestion
		container->connect(SceneStringName(draw), callable_mp(CanvasItemEditor::get_singleton(), &CanvasItemEditor::update_viewport));
		container->connect(SceneStringName(sort_children), callable_mp(CanvasItemEditor::get_singleton(), &CanvasItemEditor::update_viewport));
```

AThousandShips replied:
```suggestion
		container->connect(SceneStringName(draw), callable_mp(CanvasItemEditor::get_singleton(), &CanvasItemEditor::update_viewport));
		container->connect(SceneStringName(sort_children), callable_mp(CanvasItemEditor::get_singleton(), &CanvasItemEditor::update_viewport));
```

AThousandShips commented on scene/gui/grid_container.cpp:
```diff
@@ -217,15 +218,24 @@ void GridContainer::_notification(int p_what) {
+					cell_sizes.append(p.x);
+				}
+
+				if (row > 0 && col == 0) {
```
```suggestion
				} else if (row > 0 && col == 0) {
```
I'd say to help branching

Meorge commented:
Awesome! I know we're up close to the feature freeze deadline, but I really hope this is able to make it into 4.6 alongside/replacing the current MarginContainer visualization. Having just that one visualization for a whole release cycle might be a bit strange 😅

Calinou commented on editor/scene/gui/container_editor_plugin.cpp:
```diff
@@ -0,0 +1,290 @@
+	if (container) {
+		container->connect(SNAME("draw"), callable_mp(CanvasItemEditor::get_singleton(), &CanvasItemEditor::update_viewport));
+		container->connect(SNAME("sort_children"), callable_mp(CanvasItemEditor::get_singleton(), &CanvasItemEditor::update_viewport));
+		// minimum_size_changed is required for BoxContainer because it indirectly resorts its children instead of calling queue_resort when vertical is changed.
```
```suggestion
		// `minimum_size_changed` is required for BoxContainer, because it indirectly resorts its children instead of calling `queue_resort()` when `vertical` is changed.
```

Calinou reviewed (commented):
Tested locally, it works as expected.

Some feedback:

- The `friend class ContainerEditorPlugin;` *should* be able to be put behind `#ifdef TOOLS_ENABLED`, so that they aren't compiled in export templates. The same goes for the `cell_sizes` member which is otherwise unused.

Calinou reviewed (approved):
Tested locally (rebased on top of `master` c3e6b2c09354f804f6cf60c73f1513ab1e31ccb0), it works as expected. Code looks good to me.

Note that the way this is implemented means it's not usable at runtime, but for runtime UI debugging, a better solution may be to draw a rectangle around every Control instead (similar to `* { outline: 1px solid red; }` in CSS).

sockeye-d commented:
Could it be integrated into the rest of the debug drawing system then? How would that be possible?

AThousandShips commented on editor/scene/gui/container_editor_plugin.cpp:
```diff
@@ -0,0 +1,290 @@
+void ContainerEditorPlugin::edit(Object *p_object) {
+	if (container) {
+		container->disconnect(SNAME("draw"), callable_mp(CanvasItemEditor::get_singleton(), &CanvasItemEditor::update_viewport));
+		container->disconnect(SNAME("sort_children"), callable_mp(CanvasItemEditor::get_singleton(), &CanvasItemEditor::update_viewport));
```
```suggestion
		container->disconnect(SceneStringName(draw), callable_mp(CanvasItemEditor::get_singleton(), &CanvasItemEditor::update_viewport));
		container->disconnect(SceneStringName(sort_children), callable_mp(CanvasItemEditor::get_singleton(), &CanvasItemEditor::update_viewport));
```

sockeye-d replied:
Fixed

AThousandShips commented on editor/scene/gui/container_editor_plugin.h:
```diff
@@ -31,14 +31,26 @@
+	};
 
-	MarginContainer *margin_container = nullptr;
+	ContainerType type;
```
```suggestion
	ContainerType type = CONTAINER_TYPE_NONE;
```
Or some appropriate default

AThousandShips reviewed (approved):
Style looks good haven't tested

Calinou commented:
> Could it be integrated into the rest of the debug drawing system then? How would that be possible?

We don't have a dedicated debug drawing system for Controls yet, so we'd have to design that. The closest we have is the **Debug CanvasItem Redraw** option in the Debug menu.

KoBeWi commented on editor/scene/gui/container_editor_plugin.cpp:
```diff
@@ -0,0 +1,303 @@
+				bool draw_row_lines = (int)cell_index >= grid_container->get_columns() - 1;
+				Vector2 p1, p2;
+				if (draw_row_lines) {
+					p1 = rect.position + Vector2(0.0, cell_offset);
+					p2 = rect.position + Vector2(rect.size.x, cell_offset);
+				} else {
+					p1 = rect.position + Vector2(cell_offset, 0.0);
+					p2 = rect.position + Vector2(cell_offset, rect.size.y);
+				}
+				p_viewport_control->draw_line(xform.xform(p1), xform.xform(p2), border_color, border_width);
+			}
+		} break;
+		case CONTAINER_TYPE_CENTER: {
+			Vector2 p1, p2;
+			Vector2 center = rect.get_center();
+			double cross_width = border_width * 5.0;
+			p1 = center + Vector2(-cross_width, 0.0);
+			p2 = center + Vector2(+cross_width, 0.0);
+			p_viewport_control->draw_line(xform.xform(p1), xform.xform(p2), border_color, border_width);
+
+			p1 = center + Vector2(0.0, -cross_width);
+			p2 = center + Vector2(0.0, +cross_width);
+			p_viewport_control->draw_line(xform.xform(p1), xform.xform(p2), border_color, border_width);
+		} break;
+		case CONTAINER_TYPE_SCROLL: {
+		} break;
```
Why is this empty? 🙃

sockeye-d replied:
I was thinking about adding something for scroll containers, but decided not to. If I have any ideas in the future, I'll add them there

KoBeWi reviewed (commented):
CenterContainer with `use_top_left` still marks the center.
<img width="1421" height="620" alt="image" src="https://github.com/user-attachments/assets/782c189d-45bb-45d3-8342-98d1b914b77c" />
You could also handle PanelContainer (show the content margins).

sockeye-d commented:
I fixed the CenterContainer issues and added a new visualization for ScrollContainer that shows the expandable bounds of the children

[Screencast_20260804_041732.webm](https://github.com/user-attachments/assets/e2a9b1f0-68a9-4662-aaa7-00ee89bb69b2)

KoBeWi commented on editor/scene/gui/container_editor_plugin.h:
```diff
@@ -31,14 +31,27 @@
 
 #include "editor/plugins/editor_plugin.h"
-#include "scene/gui/margin_container.h"
+#include "scene/gui/container.h"
```
The container include is unnecessary. You can forward-declare it.

KoBeWi commented on editor/scene/gui/container_editor_plugin.cpp:
```diff
@@ -0,0 +1,325 @@
+				Vector2 p2 = p1 + Vector2(content.size.width, 0);
+				p_viewport_control->draw_line(xform.xform(p1), xform.xform(p2), border_color, border_width);
+			}
+
+			// p_viewport_control->draw_rect(xform.xform(content), border_color, false, border_width);
```
```suggestion
```

KoBeWi commented on editor/scene/gui/container_editor_plugin.cpp:
```diff
@@ -0,0 +1,325 @@
+				content.size.x = -content.size.x;
+			}
+
+			if (scroll_container->theme_cache.scrollbar_h_separation != 0 && content.size.width != rect.size.width) {
```
```suggestion
			if (scroll_container->theme_cache.scrollbar_h_separation != 0 && content.size.width < rect.size.width) {
```
Same below.

romgerman commented:
> All the TOOLS_ENABLED bits are making mess in the code. The FlowContainer is especially bad. It'd be nice to find a better way to handle it, though I guess that either means duplicate code or some major refactor :/

It really shows how non-extensible the built-in containers are.

sockeye-d commented:
@KoBeWi How don't the scroll containers show the content margins right now? I thought I made them show the bounds the child will expand to, were you thinking of something else?

KoBeWi commented:
I mean the panel margins
<img width="358" height="258" alt="image" src="https://github.com/user-attachments/assets/b5a1a5dd-8cc5-4a05-a67a-df2ff02905c8" />

sockeye-d commented:
Huh, I didn't realize ScrollContainer could have a panel. I'll try adding that
